File: SVR/svr.py
# SVR

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X_scaled = sc_X.fit_transform(X)
y_scaled = sc_y.fit_transform(y.reshape(len(y), 1)).flatten() # Reshaping the y array to a 2D array





# Training the SVR model on the whole dataset
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X_scaled, y_scaled)

# Making a single prediction
pred_scaled = regressor.predict(sc_X.transform([[6.6]]))
print(sc_y.inverse_transform(pred_scaled.reshape(1, 1)))

# Visualizing the SVR results
plt.scatter(X,y, color = 'red')
y_pred_scaled = regressor.predict(X_scaled)
y_pred = sc_y.inverse_transform(y_pred_scaled.reshape(len(y_pred_scaled), 1))
plt.plot(X, y_pred, color = 'blue')
plt.title('SVR')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()



# Visualizing the SVR results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X) + 0.1, 0.01) # choice of 0.01 instead of 0.1 step because the data is feature scaled
X_grid = X_grid.reshape((len(X_grid), 1)) # Reshaping the X_grid to a 2D array
X_grid_scaled = sc_X.transform(X_grid)
plt.scatter(X,y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('SVR')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: 5.SVR/svr.R
# SVR (Support Vector Regression)

# Importing the dataset from a CSV file named 'Position_Salaries.csv'
dataset = read.csv('Position_Salaries.csv')

# Selecting only columns 2 and 3 from the dataset
# This likely keeps only the Level and Salary columns, removing any ID column
dataset = dataset[, 2:3]

# Loading the e1071 package which contains the SVM implementation in R
library(e1071) # install.packages('e1071')

# Fitting a Support Vector Regression model to the dataset
# The formula Salary ~ . means predict Salary using all other variables (in this case, just Level)
# type = 'eps-regression' specifies epsilon-SVR algorithm (standard for regression problems)
# kernel = 'radial' uses a Radial Basis Function (RBF) kernel, which is effective for nonlinear relationships
regressor = svm(formula = Salary ~ .,
                data = dataset,
                type = 'eps-regression', # epds_regression is the default for SVR
                kernel = 'radial') # radial is the default for SVR

# Making a single prediction for a new data point (Level = 6.6)
# This predicts the salary for someone at position level 6.6
print(predict(regressor, newdata = data.frame('Level' = c(6.6))))

# Visualizing the SVR results
# Loading the ggplot2 visualization package
library(ggplot2)

# Creating a basic plot showing the SVR model and data points
ggplot() +
  # Adding red scatter points representing actual data
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  # Adding blue line showing the SVR model's predictions at each Level in the dataset
  geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
            colour = 'blue') +
  # Adding title and axis labels
  ggtitle('SVR') +
  xlab('Level') +
  ylab('Salary')

# Creating a smoother, higher-resolution visualization of the SVR model
library(ggplot2)

# Creating a sequence of Level values with smaller intervals for smoother plotting
# The sequence goes from minimum to maximum Level value with steps of 0.1
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

# Creating a more detailed plot with the finer grid
ggplot() +
  # Adding the original data points in red
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  # Adding a smoother blue curve using predictions from the model on the finer grid
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame('Level' = x_grid))),
            colour = 'blue') +
  # Adding title and axis labels
  ggtitle('SVR') +
  xlab('Level') +
  ylab('Salary')